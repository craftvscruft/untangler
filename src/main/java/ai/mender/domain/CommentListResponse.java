package ai.mender.domain;

import ai.mender.untangler.shared.TableOutput;

import java.io.PrintWriter;
import java.util.List;

public record CommentListResponse(boolean success, String message, List<CommentRec> items) {
    public void writeTableOutput(PrintWriter writer) {
        TableOutput.writeTableOutput(writer, this.items, CommentRec.class);
    }
}
