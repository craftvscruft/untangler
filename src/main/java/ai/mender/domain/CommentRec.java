package ai.mender.domain;

public record CommentRec(SourceRange range, String text) {
}
