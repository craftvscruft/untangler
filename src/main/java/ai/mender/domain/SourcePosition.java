package ai.mender.domain;

import org.apache.commons.lang3.compare.ComparableUtils;

import java.util.Comparator;

public record SourcePosition(int line, int col) implements Comparable<SourcePosition> {

    public static SourcePosition parse(String text) {
        String[] startEnd = text.split(":");
        return new SourcePosition(Integer.parseInt(startEnd[0]), Integer.parseInt(startEnd[1]));
    }

    @Override
    public String toString() {
        return String.format("%s:%s", line, col);
    }



    @Override
    public int compareTo(SourcePosition o) {
        return Comparator.comparing(SourcePosition::line).thenComparing(SourcePosition::col).compare(this, o);
    }

    public boolean isWithin(SourceRange outerRange) {
        // Inclusive
        return ComparableUtils.is(this).between(outerRange.start(), outerRange.end());
    }
}
