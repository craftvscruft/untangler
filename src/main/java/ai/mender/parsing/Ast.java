package ai.mender.parsing;

import ai.mender.domain.SourceRange;
import com.google.common.base.Objects;

import java.util.List;
import java.util.Optional;
import java.util.stream.IntStream;
import java.util.stream.Stream;

public record Ast(String tag, Ast parent, List<Ast> children, SourceRange range){

    public void addChild(Ast child) {
        children().add(child);
    }
    public boolean isChildless() {
        return children.isEmpty();
    }

    public boolean allChildrenAreChildless() {
        return children.stream().allMatch(Ast::isChildless);
    }

    public Stream<Ast> reverseChildrenStream() {
        var children = children();
        IntStream reverseIndices = IntStream.range(0, children.size()).map(i -> children.size() - i - 1);
        return reverseIndices.mapToObj(children::get);
    }

    public Optional<Ast> getFirstChildByTag(String tag) {
        return children().stream().filter((child) -> tag.equals(child.tag())).findFirst();
    }

    public List<String> childTags() {
        return children.stream().map(Ast::tag).toList();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Ast ast = (Ast) o;
        return Objects.equal(tag, ast.tag) && Objects.equal(children, ast.children) && Objects.equal(range, ast.range);
    }

    @Override
    public int hashCode() {
        // Do not include parents.
        return Objects.hashCode(tag, children, range);
    }

    @Override
    public String toString() {
        return "Ast{" +
                "tag='" + tag + '\'' +
                ", children=" + childTags() +
                ", range=" + range +
                '}';
    }
}
