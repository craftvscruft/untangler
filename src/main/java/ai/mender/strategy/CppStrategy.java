package ai.mender.strategy;

import ai.mender.Language;
import ai.mender.domain.FunctionRec;
import ai.mender.parsing.CharsetUtils;
import ai.mender.parsing.CppFunctionDefinitionNode;
import antlrgen.cpp14.CPP14Parser;
import antlrgen.cpp14.CPP14ParserBaseListener;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.tree.ParseTreeWalker;

import java.io.File;
import java.io.IOException;
import java.util.List;

public class CppStrategy implements LanguageStrategy {

    @Override
    public void collectFunctions(
            File file, List<FunctionRec> items, boolean throwOnParseError) {
        CPP14ParserBaseListener listener =
                new CPP14ParserBaseListener() {

                    @Override
                    public void enterFunctionDefinition(CPP14Parser.FunctionDefinitionContext ctx) {
                        items.add(new CppFunctionDefinitionNode(ctx).toFunctionRec());
                    }
                };

        CPP14Parser.TranslationUnitContext tree = parseTree(file, throwOnParseError);
        ParseTreeWalker.DEFAULT.walk(listener, tree);
    }

    private static CPP14Parser.TranslationUnitContext parseTree(File file, boolean throwOnParseError) {
        CharStream inputStream =
                null;
        try {
            inputStream = CharStreams.fromFileName(
                    file.getAbsolutePath(), CharsetUtils.detectFileCharset(file));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        CPP14Parser.TranslationUnitContext tree =
                Language.parseProgram(inputStream, throwOnParseError);
        return tree;
    }
}
