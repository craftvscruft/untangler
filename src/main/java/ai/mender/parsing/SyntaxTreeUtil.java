package ai.mender.parsing;

import ai.mender.untangler.shared.ISourceFile;
import ai.mender.strategy.TreeBuilder;
import ai.mender.untangler.shared.response.SourcePosition;
import ai.mender.untangler.shared.response.SourceRange;
import ai.mender.untangler.shared.response.SourceText;
import antlrgen.cpp14.CPP14Parser;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.misc.Interval;
import org.antlr.v4.runtime.tree.TerminalNode;

import java.io.IOException;
import java.util.List;
import java.util.Stack;
import java.util.stream.Collectors;

public class SyntaxTreeUtil {
    private SyntaxTreeUtil() {

    }
    public static String getTextIncludingWhitespace(ParserRuleContext ctx) {
        if (null == ctx) {
            return "";
        }
        int a = ctx.start.getStartIndex();
        int b = ctx.stop.getStopIndex();
        return ctx.start.getInputStream().getText(new Interval(a, b));
    }

    public static String getTextIncludingWhitespace(TerminalNode ctx) {
        if (null == ctx) {
            return "";
        }
        return ctx.getText();
    }


    public static SourceRange nodeToSourceRange(ParserRuleContext antlrNode) {
        int startColumn = antlrNode.start.getCharPositionInLine() + 1;
        SourcePosition start = new SourcePosition(antlrNode.start.getLine(), startColumn);
        int stopColumn = antlrNode.stop.getCharPositionInLine() + antlrNode.stop.getText().length() + 1;
        SourcePosition end = new SourcePosition(antlrNode.stop.getLine(), stopColumn);
        return new SourceRange(start, end);
    }

    public static SourceRange nodeToSourceRange(Token token) {
        int startColumn = token.getCharPositionInLine() + 1;
        int startLine = token.getLine();
        SourcePosition start = new SourcePosition(startLine, startColumn);
        List<String> lines = token.getText().lines().toList();
        int lastLineLength = lines.get(lines.size() - 1).length();
        int endLine = startLine + lines.size() - 1;
        int stopColumn = lines.size() > 1 ? (lastLineLength + 1) : startColumn + lastLineLength;
        SourcePosition end = new SourcePosition(endLine, stopColumn);
        return new SourceRange(start, end);
    }

    public static String astToSexpr(Ast tree) {
        return astToSexpr(tree, false);
    }

    public static String astToSexprWithRanges(Ast ast) {
        return astToSexpr(ast, true);
    }

    private static String astToSexpr(Ast tree, boolean includeRanges) {
        final String INDENT = "  ";
        StringBuilder stringBuilder = new StringBuilder();
        Stack<TreeBuilder.AstDepth> openNodes = new Stack<>();
        openNodes.add(new TreeBuilder.AstDepth(tree, 0));

        int prevDepth = 0;
        while (!openNodes.isEmpty()) {
            TreeBuilder.AstDepth nodeDepth = openNodes.pop();
            Ast node = nodeDepth.node();
            int depth = nodeDepth.depth();
            if (depth < prevDepth) {
                stringBuilder.append(")".repeat(prevDepth - depth));
            }
            if (depth > 0) {
                stringBuilder.append("\n").append(INDENT.repeat(depth));
            }
            if (node.isChildless()) {
                stringBuilder.append(renderNodeTag(node, includeRanges));
            } else {
                stringBuilder.append("(");
                stringBuilder.append(renderNodeTag(node, includeRanges));

                if (node.allChildrenAreChildless()) {
                    stringBuilder.append(" ").append(node.children().stream().map(Ast::tag).collect(Collectors.joining(" ")));
                    stringBuilder.append(")");
                } else {
                    node.reverseChildrenStream()
                            .map(child -> new TreeBuilder.AstDepth(child, depth+1))
                            .forEachOrdered(openNodes::push);
                }
            }
            prevDepth = depth;
        }
        stringBuilder.append(")".repeat(prevDepth));
        return stringBuilder.toString();
    }

    private static String renderNodeTag(Ast node, boolean includeRange) {
        if (includeRange) {
            return node.tag() + (node.range() == null ? "[missing]" : node.range().toString());
        }
        return node.tag();
    }


    public static SourceText nodeToSourceText(CPP14Parser.InitDeclaratorContext ctx) {
        return new SourceText(
                getTextIncludingWhitespace(ctx),
                nodeToSourceRange(ctx));
    }

    public static CharStream fileToCharStream(ISourceFile file) {
        try {
            return CharStreams.fromReader(file.getReader(), file.getName());
        } catch (IOException e) {
            throw new RuntimeException(e.getMessage(), e);
        }
    }
}
