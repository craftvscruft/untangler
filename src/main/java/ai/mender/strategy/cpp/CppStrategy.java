package ai.mender.strategy.cpp;

import ai.mender.domain.*;
import ai.mender.parsing.*;
import ai.mender.strategy.ISourceFile;
import ai.mender.strategy.LanguageStrategy;
import ai.mender.strategy.TopLevelNode;
import ai.mender.strategy.TreeBuilder;
import antlrgen.cpp14.CPP14Lexer;
import antlrgen.cpp14.CPP14Parser;
import antlrgen.cpp14.CPP14ParserBaseListener;
import com.google.common.annotations.VisibleForTesting;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.Token;
import org.antlr.v4.runtime.tree.ParseTreeWalker;
import org.apache.commons.lang3.function.Consumers;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.function.Consumer;

public class CppStrategy implements LanguageStrategy {
    final Logger LOG = LoggerFactory.getLogger(CppStrategy.class);

    @VisibleForTesting
    public static Ast getTransformedAst(Ast ast) {
        AstTransformer transformer = new AstTransformer();
        transformer.registerTrackingListener("fn");
        transformer.registerTrackingListener("fnBody");
        transformer.getListenerRegistry().listenForTag("idExpr",
                node -> {
                    if (transformer.hasAncestorWithTag("fn") &&
                            !transformer.parentTagsStoppingAt("fn").contains("fnBody")) {

                        node.children().forEach(child -> {
                            transformer.treeBuilder.setCurrentRange(child.range());
                            transformer.treeBuilder.addFirstOrAccumulate("name", "params", child.tag());
                        });
                        transformer.treeBuilder.setCurrentRange(node.range());

                    } else if (transformer.hasAncestorWithTag("simpleDeclWithSpec") &&
                            !transformer.hasAncestorWithTag("initializer")) {
                        node.children().forEach(child -> {
                                transformer.treeBuilder.setCurrentRange(child.range());
                                transformer.treeBuilder.accumulate("decl", child.tag());
                        });
                        transformer.treeBuilder.setCurrentRange(node.range());
                    } else {
                        node.children().forEach(child -> {
                            transformer.treeBuilder.setCurrentRange(child.range());
                            transformer.treeBuilder.accumulate("ref", child.tag());
                        });
                        transformer.treeBuilder.setCurrentRange(node.range());
                    }
                }, node -> {});


        return transformer.transform(ast);
    }

    @VisibleForTesting
    public static Ast parseAst(CppTopLevelNode cppTopLevelNode) {
        TreeBuilder treeBuilder = new TreeBuilder();
        treeBuilder.setCurrentRange(cppTopLevelNode.sourceRange());
        ListenerRegistry listenerRegistry = new ListenerRegistry();
        listenerRegistry.register(CPP14Parser.FunctionDefinitionContext.class,
                ctx -> treeBuilder.enter("fn"),
                ctx -> treeBuilder.exit());
        listenerRegistry.register(CPP14Parser.FunctionBodyContext.class,
                ctx -> treeBuilder.enter("fnBody"),
                ctx -> treeBuilder.exit());
        listenerRegistry.register(CPP14Parser.InitializerContext.class,
                ctx -> treeBuilder.enter("initializer"),
                ctx -> treeBuilder.exit());
        listenerRegistry.register(CPP14Parser.SimpleDeclarationContext.class,
                ctx -> treeBuilder.enter(ctx.declSpecifierSeq() == null ? "simpleDecl" : "simpleDeclWithSpec"),
                ctx -> treeBuilder.exit());
        listenerRegistry.register(CPP14Parser.IdExpressionContext.class,
                ctx -> treeBuilder.accumulate("idExpr", ctx.getText()),
                Consumers.nop());

        var listenerNew = new CPP14ParserBaseListener() {
            @Override
            public void enterEveryRule(ParserRuleContext ctx) {
                treeBuilder.setCurrentRange(SyntaxTreeUtil.nodeToSourceRange(ctx));
                listenerRegistry.acceptEnter(ctx);
            }

            @Override
            public void exitEveryRule(ParserRuleContext ctx) {
                listenerRegistry.acceptExit(ctx);
            }
        };
        ParseTreeWalker.DEFAULT.walk(listenerNew, cppTopLevelNode.getTree());


        return treeBuilder.root();
    }

    @Override
    public CppTopLevelNode parseTopLevel(ISourceFile sourceFile) {
        boolean throwOnParseError = false;
        CPP14Lexer lexer = new CPP14Lexer(sourceFile.getCharStream());
        if (throwOnParseError) {
            lexer.addErrorListener(ThrowingErrorListener.INSTANCE);
        }

        CommonTokenStream commonTokenStream = new CommonTokenStream(lexer);
        CPP14Parser parser = new CPP14Parser(commonTokenStream);
        if (throwOnParseError) {
            parser.addErrorListener(ThrowingErrorListener.INSTANCE);
        }
        CPP14Parser.TranslationUnitContext tree =
                parser.translationUnit();
        return new CppTopLevelNode(tree, parser);
    }

    public ReferencesResponse references(TopLevelNode root, String nameToResolve) {
//        LOG.debug("References");

        CppTopLevelNode cppRoot = (CppTopLevelNode) root;
        Ast tree = getTransformedAst(parseAst(cppRoot));
        Scope scope = new Scope(nameToResolve);

        AstWalker astWalker = new AstWalker();
//        astWalker
        /* Example:

(root
  (decl global1)
  (fn
    (name foo)
    (params arg1 arg1)
    (fnBody
      (decl s)
      (ref currentPoint calculateBezierPoint t calculateBezierPoint u global1 arg1))))
         */
        astWalker.getListenerRegistry().listenForTag("decl",
                (node)-> node.children()
                        .forEach(scope::addDeclaration),
                (node)->{});
        astWalker.getListenerRegistry().listenForTag("params",
                (node)-> node.children()
                        .forEach(scope::addDeclaration),
                (node)->{});
        astWalker.getListenerRegistry().listenForTag("fn",
                (fnNode)-> {
                    Optional<Ast> maybeNameNode = fnNode.getFirstChildByTag("name");
                    // Define fn name in the scope
                    maybeNameNode.ifPresent(nameNode -> {
                        List<String> nameTags = nameNode.childTags();
                        if (!nameTags.isEmpty()) {
                            // Warn if size not 1?
                            scope.addDeclaration(nameNode.children().get(0));
                        }
                    });
                    scope.push();
                },
                (node)-> scope.pop());
        astWalker.getListenerRegistry().listenForTag("ref",
                (node)-> node.children()
                        .forEach(refNode -> scope.addRef(refNode)),
                (node)->{});
        astWalker.walk(tree);
        scope.resolveUnknownFromCurrentScope();
        return scope.getReferences();
    }

    @Override
    public void forEachComment(ISourceFile sourceFile, Consumer<CommentRec> consumer) {
        CPP14Lexer lexer = new CPP14Lexer(sourceFile.getCharStream());
        CommonTokenStream commonTokenStream = new CommonTokenStream(lexer, Token.HIDDEN_CHANNEL);
        Token token;
        while ((token = commonTokenStream.getTokenSource().nextToken()).getType()!= Token.EOF) {
            if (token.getType() == CPP14Lexer.BlockComment || token.getType() == CPP14Lexer.LineComment) {
                consumer.accept(new CommentRec(SyntaxTreeUtil.nodeToSourceRange(token), token.getText()));
            }
        }
    }

    /**
     * Run consumer for every variable declaration statement.
     * Included:
     *   * int i = 0;
     *   * int a,b,c;
     * Not included:
     *   int foo(int i) {}
     *   for (int i = 0;;) {}
     * @param root
     * @param consumer
     */
    public void forEachVarDeclarationStatement(TopLevelNode root, Consumer<DeclarationRec> consumer) {
        /* From the Grammar:
        simpleDeclaration:
            declSpecifierSeq? initDeclaratorList? Semi
            | attributeSpecifierSeq declSpecifierSeq? initDeclaratorList Semi;

        forInitStatement: expressionStatement | simpleDeclaration;
         */
        CppTopLevelNode cppRoot = (CppTopLevelNode) root;
        List<DeclarationRec> decls = new ArrayList<>();

        var listener = new CPP14ParserBaseListener() {
            boolean isInForInit = false;
            @Override
            public void enterForInitStatement(CPP14Parser.ForInitStatementContext ctx) {
                isInForInit = true;
            }

            @Override
            public void exitForInitStatement(CPP14Parser.ForInitStatementContext ctx) {
                isInForInit = false;
            }

            @Override
            public void enterSimpleDeclaration(CPP14Parser.SimpleDeclarationContext ctx) {
                // We don't know how to handle whatever attributes are.
                if (ctx.attributeSpecifierSeq() == null && !isInForInit) {
                    SourceRange range = SyntaxTreeUtil.nodeToSourceRange(ctx);
                    String specifierText = SyntaxTreeUtil.getTextIncludingWhitespace(ctx.declSpecifierSeq());
                    decls.add(new DeclarationRec(range, specifierText, new ArrayList<>()));
                }
            }

            @Override
            public void enterInitDeclarator(CPP14Parser.InitDeclaratorContext ctx) {
                if (!decls.isEmpty() && !isInForInit) {
                    DeclarationRec declaration = decls.get(decls.size() - 1);
                    SourceText sourceText = SyntaxTreeUtil.nodeToSourceText(ctx);
                    System.out.println(sourceText);
                    if (sourceText.range().isWithin(declaration.range())) {
                        declaration.declarators().add(sourceText);
                    }
                }
            }
        };
        ParseTreeWalker.DEFAULT.walk(listener, cppRoot.getTree());
        decls.forEach(consumer::accept);
    }

}
