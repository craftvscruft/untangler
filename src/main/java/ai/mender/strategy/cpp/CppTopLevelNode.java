package ai.mender.strategy.cpp;

import ai.mender.domain.FunctionRec;
import ai.mender.strategy.TopLevelNode;
import antlrgen.cpp14.CPP14Parser;
import antlrgen.cpp14.CPP14ParserBaseListener;
import org.antlr.v4.runtime.tree.ParseTreeWalker;

import java.util.List;

public class CppTopLevelNode implements TopLevelNode {
    private final CPP14Parser.TranslationUnitContext tree;

    public CppTopLevelNode(CPP14Parser.TranslationUnitContext tree) {
        this.tree = tree;
    }

    @Override
    public void collectFunctions(List<FunctionRec> items) {
        CPP14ParserBaseListener listener =
                new CPP14ParserBaseListener() {

                    @Override
                    public void enterFunctionDefinition(CPP14Parser.FunctionDefinitionContext ctx) {
                        items.add(new CppFunctionDefinitionNode(ctx).toFunctionRec());
                    }
                };

        ParseTreeWalker.DEFAULT.walk(listener, tree);
    }
}
