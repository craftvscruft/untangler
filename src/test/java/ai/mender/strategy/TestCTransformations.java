package ai.mender.strategy;

import ai.mender.domain.SourcePosition;
import ai.mender.parsing.Ast;
import ai.mender.parsing.SyntaxTreeUtil;
import ai.mender.strategy.cpp.CppStrategy;
import ai.mender.strategy.cpp.CppTopLevelNode;
import org.apache.commons.lang3.StringUtils;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public class TestCTransformations {
    @Test
    public void transformedAstGroupsVariableDeclarationsAndReferences() {
        String code = """
            double global1;
            void foo(double arg1, double arg2) {
              double s = currentPoint = calculateBezierPoint(t) - calculateBezierPoint(u);
              global1 = arg1;
            }""";

        // Final pass should look something like this:
        // (root
        //  (decl global1)
        //  (fn
        //    (name foo)
        //    (params arg1 arg1)
        //    (fnBody
        //      (decl s)
        //      (ref currentPoint calculateBezierPoint t calculateBezierPoint u global1 arg1))))
        //

//        (root
//                (fn
//                        (name arg1)
//        (params arg1)
//        (fnBody
//                (decl initializer)
//        (ref calculateBezierPoint t calculateBezierPoint u simpleDecl initializer))))
        SourceFile.StringSourceFile sourceFile = new SourceFile.StringSourceFile("main.c", code);
        CppTopLevelNode topLevelNode = new CppStrategy().parseTopLevel(sourceFile);
        List<String> astPhases = topLevelNode.exposeAstPasses();
        Approvals.verify(StringUtils.join(astPhases, "\n====PASS===\n"));
    }

    @Test
    public void astTransformationKeepsSourceRangesOfFunctionName() {
        String code = """
            double global1;
            void foo(double arg1, double arg2) {
              double s = currentPoint = calculateBezierPoint(t) - calculateBezierPoint(u);
              global1 = arg1;
            }""";
        SourceFile.StringSourceFile sourceFile = new SourceFile.StringSourceFile("main.c", code);
        CppTopLevelNode topLevelNode = new CppStrategy().parseTopLevel(sourceFile);
        Ast ast = CppStrategy.getTransformedAst(CppStrategy.parseAst(topLevelNode));
        Optional<Ast> fnNode = ast.getFirstChildByTag("fn");
        Assertions.assertFalse(fnNode.isEmpty());
        Optional<Ast> nameNode = fnNode.get().getFirstChildByTag("name");
        Assertions.assertFalse(nameNode.isEmpty());
        Assertions.assertEquals(new SourcePosition(2, 6), nameNode.get().range().start());
        Assertions.assertEquals(new SourcePosition(2, 9), nameNode.get().range().end());
    }

    @Test
    public void astTransformationKeepsSourceRangesOfParamName() {
        String code = """
            void foo(double arg1, double arg2) {
            }""";
        SourceFile.StringSourceFile sourceFile = new SourceFile.StringSourceFile("main.c", code);
        CppTopLevelNode topLevelNode = new CppStrategy().parseTopLevel(sourceFile);
        Ast ast = CppStrategy.getTransformedAst(CppStrategy.parseAst(topLevelNode));
        Optional<Ast> fnNode = ast.getFirstChildByTag("fn");
        Assertions.assertFalse(fnNode.isEmpty());
        Optional<Ast> params = fnNode.get().getFirstChildByTag("params");
        Assertions.assertFalse(params.isEmpty());
        List<Ast> paramNodes = params.get().children();
        Assertions.assertEquals(2, paramNodes.size(), paramNodes.toString());
        Assertions.assertEquals(new SourcePosition(1, 17), paramNodes.get(0).range().start());
        Assertions.assertEquals(new SourcePosition(1, 21), paramNodes.get(0).range().end());
        Assertions.assertEquals(new SourcePosition(1, 30), paramNodes.get(1).range().start());
        Assertions.assertEquals(new SourcePosition(1, 34), paramNodes.get(1).range().end());
    }

    @Test
    public void astTransformationKeepsSourceRangesOfLocalVars() {
        String code = """
            void foo(double arg1, double arg2) {
              double s = currentPoint = calculateBezierPoint(t) - calculateBezierPoint(u);
              global1 = arg1;
            }""";
        SourceFile.StringSourceFile sourceFile = new SourceFile.StringSourceFile("main.c", code);
        CppTopLevelNode topLevelNode = new CppStrategy().parseTopLevel(sourceFile);
        Ast ast = CppStrategy.getTransformedAst(CppStrategy.parseAst(topLevelNode));

        Optional<Ast> fnNode = ast.getFirstChildByTag("fn");
        Assertions.assertFalse(fnNode.isEmpty());
        Optional<Ast> fnBody = fnNode.get().getFirstChildByTag("fnBody");
        Assertions.assertFalse(fnBody.isEmpty());
        Optional<Ast> declNode = fnBody.get().getFirstChildByTag("decl");
        Optional<Ast> refNode = fnBody.get().getFirstChildByTag("ref");
        Assertions.assertFalse(declNode.isEmpty());
        Assertions.assertFalse(declNode.get().children().isEmpty());
        Ast declaredVarNode = declNode.get().children().get(0);
        Assertions.assertFalse(refNode.isEmpty());

        Assertions.assertEquals("s", declaredVarNode.tag());
        Assertions.assertEquals(new SourcePosition(2, 10), declaredVarNode.range().start());
        Assertions.assertEquals(new SourcePosition(2, 11), declaredVarNode.range().end());
        String[] refNames = "currentPoint,calculateBezierPoint,t,calculateBezierPoint,u,global1,arg1".split(",");
        Assertions.assertEquals(Arrays.asList(refNames), refNode.get().childTags());
        Ast firstRef = refNode.get().children().get(0);
        Ast secondRef = refNode.get().children().get(1);
        System.out.println(SyntaxTreeUtil.astToSexprWithRanges(ast));
        Assertions.assertEquals(new SourcePosition(2,14), firstRef.range().start());
        Assertions.assertEquals(new SourcePosition(2,14 + "currentPoint".length()), firstRef.range().end());

        Assertions.assertEquals(new SourcePosition(2,29), secondRef.range().start());
        Assertions.assertEquals(new SourcePosition(2,29 + "calculateBezierPoint".length()), secondRef.range().end());
    }
}
