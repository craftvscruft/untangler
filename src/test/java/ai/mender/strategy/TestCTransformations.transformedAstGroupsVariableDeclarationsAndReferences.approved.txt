(translationUnit (declarationseq (declaration (blockDeclaration (simpleDeclaration (declSpecifierSeq
(declSpecifier (typeSpecifier (trailingTypeSpecifier (simpleTypeSpecifier double)))))
(initDeclaratorList (initDeclarator (declarator (pointerDeclarator (noPointerDeclarator
(declaratorid (idExpression (unqualifiedId global1)))))))) ;))) (declaration (functionDefinition
(declSpecifierSeq (declSpecifier (typeSpecifier (trailingTypeSpecifier (simpleTypeSpecifier
void))))) (declarator (pointerDeclarator (noPointerDeclarator (noPointerDeclarator (declaratorid
(idExpression (unqualifiedId foo)))) (parametersAndQualifiers ( (parameterDeclarationClause
(parameterDeclarationList (parameterDeclaration (declSpecifierSeq (declSpecifier (typeSpecifier
(trailingTypeSpecifier (simpleTypeSpecifier double))))) (declarator (pointerDeclarator
(noPointerDeclarator (declaratorid (idExpression (unqualifiedId arg1))))))) , (parameterDeclaration
(declSpecifierSeq (declSpecifier (typeSpecifier (trailingTypeSpecifier (simpleTypeSpecifier
double))))) (declarator (pointerDeclarator (noPointerDeclarator (declaratorid (idExpression
(unqualifiedId arg2))))))))) ))))) (functionBody (compoundStatement { (statementSeq (statement
(declarationStatement (blockDeclaration (simpleDeclaration (declSpecifierSeq (declSpecifier
(typeSpecifier (trailingTypeSpecifier (simpleTypeSpecifier double))))) (initDeclaratorList
(initDeclarator (declarator (pointerDeclarator (noPointerDeclarator (declaratorid (idExpression
(unqualifiedId s)))))) (initializer (braceOrEqualInitializer = (initializerClause
(assignmentExpression (logicalOrExpression (logicalAndExpression (inclusiveOrExpression
(exclusiveOrExpression (andExpression (equalityExpression (relationalExpression (shiftExpression
(additiveExpression (multiplicativeExpression (pointerMemberExpression (castExpression
(unaryExpression (postfixExpression (primaryExpression (idExpression (unqualifiedId
currentPoint))))))))))))))))) (assignmentOperator =) (initializerClause (assignmentExpression
(conditionalExpression (logicalOrExpression (logicalAndExpression (inclusiveOrExpression
(exclusiveOrExpression (andExpression (equalityExpression (relationalExpression (shiftExpression
(additiveExpression (multiplicativeExpression (pointerMemberExpression (castExpression
(unaryExpression (postfixExpression (postfixExpression (primaryExpression (idExpression
(unqualifiedId calculateBezierPoint)))) ( (expressionList (initializerList (initializerClause
(assignmentExpression (conditionalExpression (logicalOrExpression (logicalAndExpression
(inclusiveOrExpression (exclusiveOrExpression (andExpression (equalityExpression
(relationalExpression (shiftExpression (additiveExpression (multiplicativeExpression
(pointerMemberExpression (castExpression (unaryExpression (postfixExpression (primaryExpression
(idExpression (unqualifiedId t)))))))))))))))))))))) )))))) - (multiplicativeExpression
(pointerMemberExpression (castExpression (unaryExpression (postfixExpression (postfixExpression
(primaryExpression (idExpression (unqualifiedId calculateBezierPoint)))) ( (expressionList
(initializerList (initializerClause (assignmentExpression (conditionalExpression
(logicalOrExpression (logicalAndExpression (inclusiveOrExpression (exclusiveOrExpression
(andExpression (equalityExpression (relationalExpression (shiftExpression (additiveExpression
(multiplicativeExpression (pointerMemberExpression (castExpression (unaryExpression
(postfixExpression (primaryExpression (idExpression (unqualifiedId u))))))))))))))))))))))
)))))))))))))))))))))))) ;)))) (statement (declarationStatement (blockDeclaration (simpleDeclaration
(initDeclaratorList (initDeclarator (declarator (pointerDeclarator (noPointerDeclarator
(declaratorid (idExpression (unqualifiedId global1)))))) (initializer (braceOrEqualInitializer =
(initializerClause (assignmentExpression (conditionalExpression (logicalOrExpression
(logicalAndExpression (inclusiveOrExpression (exclusiveOrExpression (andExpression
(equalityExpression (relationalExpression (shiftExpression (additiveExpression
(multiplicativeExpression (pointerMemberExpression (castExpression (unaryExpression
(postfixExpression (primaryExpression (idExpression (unqualifiedId arg1))))))))))))))))))))))))
;))))) }))))) <EOF>)
====PASS===
(root
  (simpleDeclWithSpec
    (idExpr global1))
  (fn
    (idExpr foo arg1 arg2)
    (fnBody
      (simpleDeclWithSpec
        (idExpr s)
        (initializer
          (idExpr currentPoint calculateBezierPoint t calculateBezierPoint u)))
      (simpleDecl
        (idExpr global1)
        (initializer
          (idExpr arg1))))))
====PASS===
(root
  (decl global1)
  (fn
    (name foo)
    (params arg1 arg2)
    (fnBody
      (decl s)
      (ref currentPoint calculateBezierPoint t calculateBezierPoint u global1 arg1))))